File: modules/mono/build_scripts/mono_reg_utils.py
import os
import platform

from compat import decode_utf8

if os.name == "nt":
    import sys

    if sys.version_info < (3,):
        import _winreg as winreg
    else:
        import winreg


def _reg_open_key(key, subkey):
    try:
        return winreg.OpenKey(key, subkey)
    except (WindowsError, OSError):
        if platform.architecture()[0] == "32bit":
            bitness_sam = winreg.KEY_WOW64_64KEY
        else:
            bitness_sam = winreg.KEY_WOW64_32KEY
        return winreg.OpenKey(key, subkey, 0, winreg.KEY_READ | bitness_sam)


def _reg_open_key_bits(key, subkey, bits):
    sam = winreg.KEY_READ

    if platform.architecture()[0] == "32bit":
        if bits == "64":
            # Force 32bit process to search in 64bit registry
            sam |= winreg.KEY_WOW64_64KEY
    else:
        if bits == "32":
            # Force 64bit process to search in 32bit registry
            sam |= winreg.KEY_WOW64_32KEY

    return winreg.OpenKey(key, subkey, 0, sam)


def _find_mono_in_reg(subkey, bits):
    try:
        with _reg_open_key_bits(winreg.HKEY_LOCAL_MACHINE, subkey, bits) as hKey:
            value = winreg.QueryValueEx(hKey, "SdkInstallRoot")[0]
            return value
    except (WindowsError, OSError):
        return None


def _find_mono_in_reg_old(subkey, bits):
    try:
        with _reg_open_key_bits(winreg.HKEY_LOCAL_MACHINE, subkey, bits) as hKey:
            default_clr = winreg.QueryValueEx(hKey, "DefaultCLR")[0]
            if default_clr:
                return _find_mono_in_reg(subkey + "\\" + default_clr, bits)
            return None
    except (WindowsError, EnvironmentError):
        return None


def find_mono_root_dir(bits):
    root_dir = _find_mono_in_reg(r"SOFTWARE\Mono", bits)
    if root_dir is not None:
        return str(root_dir)
    root_dir = _find_mono_in_reg_old(r"SOFTWARE\Novell\Mono", bits)
    if root_dir is not None:
        return str(root_dir)
    return ""


def find_msbuild_tools_path_reg():
    import subprocess

    vswhere = os.getenv("PROGRAMFILES(X86)")
    if not vswhere:
        vswhere = os.getenv("PROGRAMFILES")
    vswhere += r"\Microsoft Visual Studio\Installer\vswhere.exe"

    vswhere_args = ["-latest", "-products", "*", "-requires", "Microsoft.Component.MSBuild"]

    try:
        lines = subprocess.check_output([vswhere] + vswhere_args).splitlines()

        for line in lines:
            parts = decode_utf8(line).split(":", 1)

            if len(parts) < 2 or parts[0] != "installationPath":
                continue

            val = parts[1].strip()

            if not val:
                raise ValueError("Value of `installationPath` entry is empty")

            # Since VS2019, the directory is simply named "Current"
            msbuild_dir = os.path.join(val, "MSBuild\\Current\\Bin")
            if os.path.isdir(msbuild_dir):
                return msbuild_dir

            # Directory name "15.0" is used in VS 2017
            return os.path.join(val, "MSBuild\\15.0\\Bin")

        raise ValueError("Cannot find `installationPath` entry")
    except ValueError as e:
        print("Error reading output from vswhere: " + e.message)
    except WindowsError:
        pass  # Fine, vswhere not found
    except (subprocess.CalledProcessError, OSError):
        pass

    # Try to find 14.0 in the Registry

    try:
        subkey = r"SOFTWARE\Microsoft\MSBuild\ToolsVersions\14.0"
        with _reg_open_key(winreg.HKEY_LOCAL_MACHINE, subkey) as hKey:
            value = winreg.QueryValueEx(hKey, "MSBuildToolsPath")[0]
            return value
    except (WindowsError, OSError):
        return ""

File: modules/mono/utils/mono_reg_utils.cpp
/*************************************************************************/
/*  mono_reg_utils.cpp                                                   */
/*************************************************************************/
/*                       This file is part of:                           */
/*                           GODOT ENGINE                                */
/*                      https://godotengine.org                          */
/*************************************************************************/
/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
/*                                                                       */
/* Permission is hereby granted, free of charge, to any person obtaining */
/* a copy of this software and associated documentation files (the       */
/* "Software"), to deal in the Software without restriction, including   */
/* without limitation the rights to use, copy, modify, merge, publish,   */
/* distribute, sublicense, and/or sell copies of the Software, and to    */
/* permit persons to whom the Software is furnished to do so, subject to */
/* the following conditions:                                             */
/*                                                                       */
/* The above copyright notice and this permission notice shall be        */
/* included in all copies or substantial portions of the Software.       */
/*                                                                       */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
/*************************************************************************/

#include "mono_reg_utils.h"
#include "core/os/dir_access.h"

#ifdef WINDOWS_ENABLED

#include "core/os/os.h"

#define WIN32_LEAN_AND_MEAN
#include <windows.h>

namespace MonoRegUtils {

template <int>
REGSAM bitness_sam_impl();

template <>
REGSAM bitness_sam_impl<4>() {
	return KEY_WOW64_64KEY;
}

template <>
REGSAM bitness_sam_impl<8>() {
	return KEY_WOW64_32KEY;
}

REGSAM _get_bitness_sam() {
	return bitness_sam_impl<sizeof(size_t)>();
}

LONG _RegOpenKey(HKEY hKey, LPCWSTR lpSubKey, PHKEY phkResult) {
	LONG res = RegOpenKeyExW(hKey, lpSubKey, 0, KEY_READ, phkResult);

	if (res != ERROR_SUCCESS)
		res = RegOpenKeyExW(hKey, lpSubKey, 0, KEY_READ | _get_bitness_sam(), phkResult);

	return res;
}

LONG _RegKeyQueryString(HKEY hKey, const String &p_value_name, String &r_value) {
	Vector<WCHAR> buffer;
	buffer.resize(512);
	DWORD dwBufferSize = buffer.size();

	LONG res = RegQueryValueExW(hKey, p_value_name.c_str(), 0, NULL, (LPBYTE)buffer.ptr(), &dwBufferSize);

	if (res == ERROR_MORE_DATA) {
		// dwBufferSize now contains the actual size
		buffer.resize(dwBufferSize);
		res = RegQueryValueExW(hKey, p_value_name.c_str(), 0, NULL, (LPBYTE)buffer.ptr(), &dwBufferSize);
	}

	if (res == ERROR_SUCCESS) {
		r_value = String(buffer.ptr(), buffer.size());
	} else {
		r_value = String();
	}

	return res;
}

LONG _find_mono_in_reg(const String &p_subkey, MonoRegInfo &r_info, bool p_old_reg = false) {
	HKEY hKey;
	LONG res = _RegOpenKey(HKEY_LOCAL_MACHINE, p_subkey.c_str(), &hKey);

	if (res != ERROR_SUCCESS)
		goto cleanup;

	if (!p_old_reg) {
		res = _RegKeyQueryString(hKey, "Version", r_info.version);
		if (res != ERROR_SUCCESS)
			goto cleanup;
	}

	res = _RegKeyQueryString(hKey, "SdkInstallRoot", r_info.install_root_dir);
	if (res != ERROR_SUCCESS)
		goto cleanup;

	res = _RegKeyQueryString(hKey, "FrameworkAssemblyDirectory", r_info.assembly_dir);
	if (res != ERROR_SUCCESS)
		goto cleanup;

	res = _RegKeyQueryString(hKey, "MonoConfigDir", r_info.config_dir);
	if (res != ERROR_SUCCESS)
		goto cleanup;

	if (r_info.install_root_dir.ends_with("\\"))
		r_info.bin_dir = r_info.install_root_dir + "bin";
	else
		r_info.bin_dir = r_info.install_root_dir + "\\bin";

cleanup:
	RegCloseKey(hKey);
	return res;
}

LONG _find_mono_in_reg_old(const String &p_subkey, MonoRegInfo &r_info) {
	String default_clr;

	HKEY hKey;
	LONG res = _RegOpenKey(HKEY_LOCAL_MACHINE, p_subkey.c_str(), &hKey);

	if (res != ERROR_SUCCESS)
		goto cleanup;

	res = _RegKeyQueryString(hKey, "DefaultCLR", default_clr);

	if (res == ERROR_SUCCESS && default_clr.length()) {
		r_info.version = default_clr;
		res = _find_mono_in_reg(p_subkey + "\\" + default_clr, r_info, true);
	}

cleanup:
	RegCloseKey(hKey);
	return res;
}

MonoRegInfo find_mono() {
	MonoRegInfo info;

	if (_find_mono_in_reg("Software\\Mono", info) == ERROR_SUCCESS)
		return info;

	if (_find_mono_in_reg_old("Software\\Novell\\Mono", info) == ERROR_SUCCESS)
		return info;

	return MonoRegInfo();
}

String find_msbuild_tools_path() {
	String msbuild_tools_path;

	// Try to find 15.0 with vswhere

	String vswhere_path = OS::get_singleton()->get_environment(sizeof(size_t) == 8 ? "ProgramFiles(x86)" : "ProgramFiles");
	vswhere_path += "\\Microsoft Visual Studio\\Installer\\vswhere.exe";

	List<String> vswhere_args;
	vswhere_args.push_back("-latest");
	vswhere_args.push_back("-products");
	vswhere_args.push_back("*");
	vswhere_args.push_back("-requires");
	vswhere_args.push_back("Microsoft.Component.MSBuild");

	String output;
	int exit_code;
	OS::get_singleton()->execute(vswhere_path, vswhere_args, true, NULL, &output, &exit_code);

	if (exit_code == 0) {
		Vector<String> lines = output.split("\n");

		for (int i = 0; i < lines.size(); i++) {
			const String &line = lines[i];
			int sep_idx = line.find(":");

			if (sep_idx > 0) {
				String key = line.substr(0, sep_idx); // No need to trim

				if (key == "installationPath") {
					String val = line.substr(sep_idx + 1, line.length()).strip_edges();

					ERR_BREAK(val.empty());

					if (!val.ends_with("\\")) {
						val += "\\";
					}

					// Since VS2019, the directory is simply named "Current"
					String msbuild_dir = val + "MSBuild\\Current\\Bin";
					if (DirAccess::exists(msbuild_dir)) {
						return msbuild_dir;
					}

					// Directory name "15.0" is used in VS 2017
					return val + "MSBuild\\15.0\\Bin";
				}
			}
		}
	}

	// Try to find 14.0 in the Registry

	HKEY hKey;
	LONG res = _RegOpenKey(HKEY_LOCAL_MACHINE, L"SOFTWARE\\Microsoft\\MSBuild\\ToolsVersions\\14.0", &hKey);

	if (res != ERROR_SUCCESS)
		goto cleanup;

	res = _RegKeyQueryString(hKey, "MSBuildToolsPath", msbuild_tools_path);

	if (res != ERROR_SUCCESS)
		goto cleanup;

cleanup:
	RegCloseKey(hKey);

	return msbuild_tools_path;
}
} // namespace MonoRegUtils

#endif // WINDOWS_ENABLED
